rust-src/ed25519_hd_key_derivation/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+    ir.clone_from_slice(&i[32..64]);
+
+    HdKeys {
+        key:        il,
+        chain_code: ir,
+    }
+}
+
+fn ckd_priv(keys: HdKeys, index: u32) -> HdKeys {
+    if index & HARDENED_OFFSET == 0 {
+        panic!("Received a non-hardened index!");
+    }
+
+    let mut data = vec![0u8];
+    data.extend_from_slice(&keys.key);
+    data.extend_from_slice(&index.to_be_bytes());
+
+    let mut mac = Hmac::<Sha512>::new_from_slice(&keys.chain_code).unwrap();
+    mac.update(&data);
+    let i = mac.finalize().into_bytes();
+    let mut il = [0u8; 32];
+    il.clone_from_slice(&i[0..32]);
+    let mut ir = [0u8; 32];
+    ir.clone_from_slice(&i[32..64]);
+
+    HdKeys {
+        key:        il,
+        chain_code: ir,
+    }
+}
+
+fn is_valid_path(path: &str) -> bool {
+    let path_regex = Regex::new("^m(/[0-9]+')+$").unwrap();
+    if !path_regex.is_match(path) {
+        return false;
+    }
+
+    let segments: Vec<&str> = path.split('/').collect();
+    segments
+        .iter()
+        .skip(1)
+        .map(|seg| seg.replace("'", ""))
+        .all(|seg| seg.parse::<u32>().is_ok())
+}
+
+/// Derives hierarchical deterministic keys for ed25519 according to the SLIP0010 (https://github.com/satoshilabs/slips/blob/master/slip-0010.md)
+/// specification.
+pub fn derive(path: &str, seed: &[u8]) -> Result<HdKeys, String> {
+    if !is_valid_path(path) {
+        return Err(String::from("Invalid derivation path"));
+    }
+
+    let master_key = get_master_key_from_seed(&seed);
+
+    let segments: Vec<&str> = path.split('/').collect();
+    let segments = segments
+        .iter()
+        .skip(1)
+        .map(|seg| seg.replace("'", ""))
+        .map(|seg| seg.parse::<u32>().unwrap())
+        .collect::<Vec<_>>();
+
+    let result: HdKeys = segments.iter().fold(master_key, |accum, index| {
+        ckd_priv(accum, *index + HARDENED_OFFSET)
+    });
+    Ok(result)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const TEST_VECTOR_1_SEED: &str = "000102030405060708090a0b0c0d0e0f";
+    const TEST_VECTOR_2_SEED: &str = "fffcf9f6f3f0edeae7e4e1dedbd8d5d2cfccc9c6c3c0bdbab7b4b1aeaba8a5a29f9c999693908d8a8784817e7b7875726f6c696663605d5a5754514e4b484542";
+
+    fn assert_keys(seed: &str, path: &str, chain_code: &str, private_key: &str, public_key: &str) {
+        let seed = hex::decode(seed).unwrap();
+        let keys = derive(&path, &seed).unwrap();
+        let secret_key = ed25519_dalek::SecretKey::from_bytes(&keys.key).unwrap();
+        let public_key_derived = ed25519_dalek::PublicKey::from(&secret_key).to_bytes();
+        assert_eq!(keys.chain_code.to_vec(), hex::decode(chain_code).unwrap());
+        assert_eq!(keys.key.to_vec(), hex::decode(private_key).unwrap());
+        assert_eq!(
+            public_key_derived.to_vec(),
+            hex::decode(public_key).unwrap()
+        )
+    }
+
+    #[test]
+    pub fn test_valid_path() {
+        let valid_path = "m/44'/919'";
+        assert!(is_valid_path(&valid_path));
```

Comment:
I usually find it is helpful to provide a message to be printed when the assertion fails.
```rust
        assert!(is_valid_path(&valid_path), "Path {} should be valid.", valid_path);
```
I find it easier to locate the errors when tests fail.